test/test_public_bindings.py

```diff
@@ -272,6 +276,888 @@ def test_no_new_bindings(self):
+                                             '_SingleProcessDataLoaderIter'],
+             'torch.utils.data.dataloader_experimental': ['_ThreadingDataLoader2',
+                                                          '_sharding_worker_init_fn'],
+             'torch.utils.data.datapipes.dataframe.dataframe_wrapper': ['_try_import_pandas',
+                                                                        '_with_pandas'],
+             'torch.utils.data.datapipes.iter.combining': ['_ChildDataPipe',
+                                                           '_DemultiplexerIterDataPipe',
+                                                           '_ForkerIterDataPipe'],
+             'torch.utils.tensorboard.summary': ['_calc_scale_factor',
+                                                 '_draw_single_box',
+                                                 '_get_json_config',
+                                                 '_get_tensor_summary']})
+        failure_list = []
+
+        def add_to_failure_list_if_not_in_allow_dict(modname, elem, elem_module):
+            if modname in allow_dict and elem in allow_dict[modname]:
+                return
+            failure_list.append((modname, elem, elem_module))
+
+        def test_module(modname):
+            split_strs = modname.split('.')
+            mod = sys.modules.get(modname)
+            for elem in split_strs:
+                if elem.startswith("_"):
+                    return
+
+            # verifies that each API has the correct module name and naming semantics
+            # depending on whether it's public or private
+            def looks_public(elem, modname, mod, private_api):
+                obj = getattr(mod, elem)
+                if not (isinstance(obj, Callable) or inspect.isclass(obj)):
+                    return
+                elem_module = getattr(obj, '__module__', None)
+                elem_modname_starts_with_modname = elem_module is not None and elem_module.startswith(modname)
+                if private_api:
+                    # elem's name must begin with an `_` and it's module name
+                    # should NOT start with it's current module since it's a private API
+                    if elem_modname_starts_with_modname:
+                        add_to_failure_list_if_not_in_allow_dict(modname, elem, elem_module)
+                else:
+                    # elem's name must NOT begin with an `_` and it's module name
+                    # SHOULD start with it's current module since it's a public API
+                    if elem.startswith('_') and not elem_modname_starts_with_modname:
+                        add_to_failure_list_if_not_in_allow_dict(modname, elem, elem_module)
+
+            if hasattr(modname, '__all__'):
+                public_api = mod.__all__
+                all_api = dir(modname)
+                for elem in all_api:
+                    looks_public(elem, modname, elem not in public_api)
```

Comment:
if the name doesn't start with underscore, then it's module name should not start with modname.